File: main/src/main/java/com/sedmelluq/discord/lavaplayer/source/youtube/YoutubeMixProvider.java
package com.sedmelluq.discord.lavaplayer.source.youtube;

import com.sedmelluq.discord.lavaplayer.tools.DataFormatTools;
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException;
import com.sedmelluq.discord.lavaplayer.tools.JsonBrowser;
import com.sedmelluq.discord.lavaplayer.tools.ThumbnailTools;
import com.sedmelluq.discord.lavaplayer.tools.io.HttpClientTools;
import com.sedmelluq.discord.lavaplayer.tools.io.HttpInterface;
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo;
import com.sedmelluq.discord.lavaplayer.track.BasicAudioPlaylist;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

import static com.sedmelluq.discord.lavaplayer.source.youtube.YoutubeHttpContextFilter.PBJ_PARAMETER;
import static com.sedmelluq.discord.lavaplayer.tools.FriendlyException.Severity.SUSPICIOUS;

/**
 * Handles loading of YouTube mixes.
 */
public class YoutubeMixProvider implements YoutubeMixLoader {
    /**
     * Loads tracks from mix in parallel into a playlist entry.
     *
     * @param mixId           ID of the mix
     * @param selectedVideoId Selected track, {@link AudioPlaylist#getSelectedTrack()} will return this.
     * @return Playlist of the tracks in the mix.
     */
    public AudioPlaylist load(
        HttpInterface httpInterface,
        String mixId,
        String selectedVideoId,
        Function<AudioTrackInfo, AudioTrack> trackFactory
    ) {
        String playlistTitle = "YouTube mix";
        List<AudioTrack> tracks = new ArrayList<>();

        String mixUrl = "https://www.youtube.com/watch?v=" + selectedVideoId + "&list=" + mixId + PBJ_PARAMETER;

        try (CloseableHttpResponse response = httpInterface.execute(new HttpGet(mixUrl))) {
            HttpClientTools.assertSuccessWithContent(response, "mix response");

            JsonBrowser body = JsonBrowser.parse(response.getEntity().getContent());
            JsonBrowser playlist = body.index(3).get("response")
                .get("contents")
                .get("twoColumnWatchNextResults")
                .get("playlist")
                .get("playlist");

            JsonBrowser title = playlist.get("title");

            if (!title.isNull()) {
                playlistTitle = title.text();
            }

            extractPlaylistTracks(playlist.get("contents"), tracks, trackFactory);
        } catch (IOException e) {
            throw new FriendlyException("Could not read mix page.", SUSPICIOUS, e);
        }

        if (tracks.isEmpty()) {
            throw new FriendlyException("Could not find tracks from mix.", SUSPICIOUS, null);
        }

        AudioTrack selectedTrack = findSelectedTrack(tracks, selectedVideoId);
        return new BasicAudioPlaylist(playlistTitle, tracks, selectedTrack, false);
    }

    private void extractPlaylistTracks(
        JsonBrowser browser,
        List<AudioTrack> tracks,
        Function<AudioTrackInfo, AudioTrack> trackFactory
    ) {
        for (JsonBrowser video : browser.values()) {
            JsonBrowser renderer = video.get("playlistPanelVideoRenderer");
            String title = renderer.get("title").get("simpleText").text();
            String author = renderer.get("longBylineText").get("runs").index(0).get("text").text();
            String identifier = renderer.get("videoId").text();
            String durationStr = renderer.get("lengthText").get("simpleText").text();
            long duration = DataFormatTools.parseDuration(durationStr);

            AudioTrackInfo trackInfo = new AudioTrackInfo(
                title,
                author,
                duration,
                identifier,
                false,
                "https://youtube.com/watch?v=" + identifier,
                ThumbnailTools.extractYouTube(renderer, identifier)
            );

            tracks.add(trackFactory.apply(trackInfo));
        }
    }

    private AudioTrack findSelectedTrack(List<AudioTrack> tracks, String selectedVideoId) {
        if (selectedVideoId != null) {
            for (AudioTrack track : tracks) {
                if (selectedVideoId.equals(track.getIdentifier())) {
                    return track;
                }
            }
        }

        return null;
    }
}

File: main/src/main/java/com/sedmelluq/discord/lavaplayer/source/youtube/music/YoutubeMixProvider.kt
package com.sedmelluq.discord.lavaplayer.source.youtube.music

import com.sedmelluq.discord.lavaplayer.source.youtube.YoutubeConstants
import com.sedmelluq.discord.lavaplayer.tools.DataFormatTools.durationTextToMillis
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException
import com.sedmelluq.discord.lavaplayer.tools.ThumbnailTools
import com.sedmelluq.discord.lavaplayer.tools.io.HttpClientTools
import com.sedmelluq.discord.lavaplayer.tools.io.HttpInterface
import com.sedmelluq.discord.lavaplayer.tools.json.JsonBrowser
import com.sedmelluq.discord.lavaplayer.track.*
import org.apache.http.client.methods.HttpPost
import org.apache.http.entity.StringEntity
import java.io.IOException

/**
 * Handles loading of YouTube mixes.
 */
class YoutubeMixProvider : YoutubeMixLoader {
    /**
     * Loads tracks from mix in parallel into a playlist entry.
     *
     * @param mixId ID of the mix
     * @param selectedVideoId Selected track, [AudioTrackCollection.selectedTrack] will return this.
     * @return Playlist of the tracks in the mix.
     */
    override fun load(
        httpInterface: HttpInterface,
        mixId: String,
        selectedVideoId: String?,
        trackFactory: AudioTrackFactory
    ): AudioTrackCollection {
        var playlistTitle = "YouTube mix"
        val tracks: MutableList<AudioTrack> = ArrayList()
        val post = HttpPost(YoutubeConstants.NEXT_URL)
        val payload = StringEntity(java.lang.String.format(YoutubeConstants.NEXT_PAYLOAD, selectedVideoId, mixId), "UTF-8")
        post.entity = payload

        try {
            httpInterface.execute(post).use { response ->
                HttpClientTools.assertSuccessWithContent(response, "mix response")
                val body = JsonBrowser.parse(response.entity.content)
                val playlist = body["contents"]["singleColumnWatchNextResults"]["results"]["results"]["contents"]

                val title = playlist["title"]
                if (!title.isNull) {
                    playlistTitle = title.safeText
                }

                println(body.format())
                extractPlaylistTracks(playlist["contents"], tracks, trackFactory)
            }
        } catch (e: IOException) {
            throw FriendlyException("Could not read mix page.", FriendlyException.Severity.SUSPICIOUS, e)
        }

        if (tracks.isEmpty()) {
            throw FriendlyException("Could not find tracks from mix.", FriendlyException.Severity.SUSPICIOUS, null)
        }

        val selectedTrack = findSelectedTrack(tracks, selectedVideoId)
        return BasicAudioTrackCollection(playlistTitle, AudioTrackCollectionType.Playlist, tracks, selectedTrack)
    }

    private fun extractPlaylistTracks(
        browser: JsonBrowser,
        tracks: MutableList<AudioTrack>,
        trackFactory: AudioTrackFactory
    ) {
        for (renderer in browser.values().map { it["playlistPanelVideoRenderer"] }) {
            val videoId = renderer["videoId"].text!!

            /* create the audio track. */
            val trackInfo = AudioTrackInfo(
                title = renderer["title"]["runs"][0]["text"].text!!,
                author = renderer["longBylineText"]["runs"][0]["text"].text!!,
                length = durationTextToMillis(renderer["lengthText"]["runs"][0]["text"].text!!),
                identifier = videoId,
                uri = "https://youtube.com/watch?v=$videoId",
                artworkUrl = ThumbnailTools.extractYouTube(renderer, videoId),
            )

            tracks.add(trackFactory.create(trackInfo))
        }
    }

    private fun findSelectedTrack(tracks: List<AudioTrack>, selectedVideoId: String?): AudioTrack? {
        if (selectedVideoId != null) {
            return tracks.find { it.identifier == selectedVideoId }
        }

        return null
    }
}

